File: frontend/src/services/knowledge.js
export function tokenize(text) {
  return (text.toLowerCase().match(/[a-z0-9_]+|[\u4e00-\u9fff]+/g) || []).flatMap(term =>
    /^[\u4e00-\u9fff]+$/.test(term) && term.length > 1
      ? Array.from({ length: term.length - 1 }, (_, i) => term.slice(i, i + 2)) : [term]);
}

export function searchDocuments(documents, query, limit = 6) {
  const terms = new Set(tokenize(query));
  const average = documents.reduce((n, d) => n + d.tokens.length, 0) / (documents.length || 1);
  const frequencies = new Map();
  for (const doc of documents) for (const term of new Set(doc.tokens)) {
    frequencies.set(term, (frequencies.get(term) || 0) + 1);
  }
  const hits = documents.map(({ tokens, ...doc }) => {
    const counts = new Map();
    for (const term of tokens) counts.set(term, (counts.get(term) || 0) + 1);
    let score = 0;
    for (const term of terms) {
      const tf = counts.get(term) || 0;
      if (!tf) continue;
      const df = frequencies.get(term);
      const idf = Math.log(1 + (documents.length - df + 0.5) / (df + 0.5));
      score += idf * tf * 2.5 / (tf + 1.5 * (0.25 + 0.75 * tokens.length / average));
    }
    return { ...doc, score: Math.round(score * 1e6) / 1e6, retrievalChannels: ['BM25'] };
  }).filter(hit => hit.score > 0).sort((a, b) => b.score - a.score || a.id.localeCompare(b.id));
  return { items: hits.slice(0, Math.min(20, Math.max(1, limit))),
    knowledgeBackend: 'browser-bm25', totalChunks: documents.length };
}

let corpus;
export async function searchPublicKnowledge(query, limit) {
  if (!corpus) {
    const response = await fetch(`${import.meta.env.BASE_URL}demo/knowledge.json`);
    if (!response.ok) throw new Error('知识文档加载失败，请重试。');
    corpus = (await response.json()).documents;
  }
  return searchDocuments(corpus, query, limit);
}

File: backend/traffic_agent/knowledge.py
from __future__ import annotations

import hashlib
import math
import re
from collections import Counter
from pathlib import Path

DOCS = Path(__file__).with_name("knowledge_docs")
VERSION = "public-knowledge-v1"


def tokenize(text):
    terms = re.findall(r"[a-z0-9_]+|[\u4e00-\u9fff]+", text.lower())
    return [token for term in terms for token in
            ([term] if not re.fullmatch(r"[\u4e00-\u9fff]+", term) or len(term) == 1
             else [term[i:i + 2] for i in range(len(term) - 1)])]


def load_chunks(directory=DOCS):
    chunks = []
    for path in sorted(directory.glob("*.md")):
        raw = path.read_text(encoding="utf-8")
        digest = hashlib.sha256(raw.encode("utf-8")).hexdigest()
        for section in raw.split("\n## ")[1:]:
            title, _, content = section.partition("\n")
            content = content.strip()
            chunk_hash = hashlib.sha256(f"{path.name}\n{title}\n{content}".encode()).hexdigest()
            chunks.append({"id": f"KB-{chunk_hash[:16]}", "title": title,
                           "content": content, "source": path.name, "section": title,
                           "sourceHash": digest, "version": VERSION,
                           "sourceType": "SELF_AUTHORED_DEMO",
                           "tokens": tokenize(title + " " + content)})
    return chunks


def search(query, limit=6, chunks=None):
    if not 1 <= limit <= 20:
        raise ValueError("limit must be between 1 and 20")
    documents = load_chunks() if chunks is None else chunks
    terms = set(tokenize(query))
    average = sum(len(d["tokens"]) for d in documents) / max(len(documents), 1)
    frequencies = Counter(t for d in documents for t in set(d["tokens"]))
    hits = []
    for doc in documents:
        counts = Counter(doc["tokens"])
        score = 0.0
        for term in terms:
            tf = counts[term]
            if tf:
                idf = math.log(1 + (len(documents) - frequencies[term] + .5) / (frequencies[term] + .5))
                score += idf * tf * 2.5 / (tf + 1.5 * (.25 + .75 * len(doc["tokens"]) / average))
        if score > 0:
            hits.append({**{k: v for k, v in doc.items() if k != "tokens"},
                         "score": round(score, 6), "retrievalChannels": ["BM25"]})
    hits.sort(key=lambda h: (-h["score"], h["id"]))
    return {"items": hits[:limit], "query": query, "knowledgeBackend": "local-bm25",
            "strategy": {"name": "bm25", "version": VERSION}, "totalChunks": len(documents)}


def retrieve_context(context):
    codes = sorted({str(e.get("code", "")) for c in context.get("candidates", [])
                    for e in c.get("evidence", []) if isinstance(e, dict) and e.get("code")})
    return search(" ".join(codes) or "调查候选 证据边界", limit=3)
